### PR TITLE
WIP: Update keycloak gem, add token class

### PR DESCRIPTION
**Pull request by samlionheart**
Ticket: https://trello.com/c/dpJZjxOi

WIP -  Needs Travis testing, as well as tests on the Token class.

Updated keycloak gem to 3.0.0

Add token class, that replaces the need for default_call in keycloak_client.rb

Token class checks if a token exists and if it has expired, generates a new token accordingly.

No longer logs out after each request to keycloak.

### CODE REVIEW

**Comment by slorek** [comment]
Looking good so far 👍

File: .gitignore [hunk 1]
```diff
@@ -144,11 +144,15 @@ $RECYCLE.BIN/
...
 *.log
 
 # Redis snapshots. See https://redis.io/topics/persistence
 dump.rdb
+
+.env.development
+.env.test
+.env
+
+.ruby-gemset
```
Comment (StupidCodeFactory) [inline .gitignore]: Don't use `gemset` it makes it makes redundant requests to rubygems.org which is kindly maintained by donations. You might reinstall gems you already have on your computer. Also bundler makes it gemsets obsoltes

File: .gitignore [hunk 2]
```diff
@@ -144,11 +144,15 @@ $RECYCLE.BIN/
...
+.ruby-version
```
Comment (StupidCodeFactory) [inline .gitignore]: you definitely want the `.ruby-version` to checked in

File: shared-web/config/initializers/shared_rails_config.rb [hunk 3]
```diff
@@ -6,4 +6,4 @@
 
 Rails.application.config.action_view.form_with_generates_ids = true
 
-ActionView::Base.send :include, Shared::Web::ComponentsHelper
+ActionView::Base.include Shared::Web::ComponentsHelper
```
Comment (StupidCodeFactory) [inline shared-web/config/initializers/shared_rails_config.rb]: 👍

File: shared-web/Gemfile.lock [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+  govuk-lint (= 3.11.2)
+  rspec
+  rubocop (= 0.71.0)
+  rubocop-performance
+  sass-rails (~> 5.0)
+  shared-web!
+  webmock
+
+BUNDLED WITH
+   1.17.3
```
Comment (StupidCodeFactory) [inline shared-web/Gemfile.lock]: can you update your bundler version to 2.0.2?

File: shared-web/Rakefile [hunk 5]
```diff
@@ -14,19 +14,9 @@ RDoc::Task.new(:rdoc) do |rdoc|
   rdoc.rdoc_files.include('lib/**/*.rb')
 end
 
-APP_RAKEFILE = File.expand_path("test/dummy/Rakefile", __dir__)
-load 'rails/tasks/engine.rake'
+# APP_RAKEFILE = File.expand_path("Rakefile", __dir__)
+# load 'rails/tasks/engine.rake'
```
Comment (StupidCodeFactory) [inline shared-web/Rakefile]: ✂️

File: shared-web/app/clients/shared/web/token.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            response.return!
+          end
+        end
+        token
+      end
+
+      def payload
+        { 'client_id' => client_id,
+        'client_secret' => secret,
+        'grant_type' => 'client_credentials' }
```
Comment (StupidCodeFactory) [inline shared-web/app/clients/shared/web/token.rb]: indentation

File: shared-web/app/clients/shared/web/token.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      end
+
+      def client_id
+        ENV.fetch("KEYCLOAK_CLIENT_ID")
+      end
+
+      def secret
+        ENV.fetch("KEYCLOAK_CLIENT_SECRET")
+      end
```
Comment (StupidCodeFactory) [inline shared-web/app/clients/shared/web/token.rb]: Could we have those two in `config/application.rb`
```ruby
config.x.keycload.client_id = ENV.fetch("KEYCLOAK_CLIENT_ID")
config.x.keycload.client_secret = ENV.fetch("KEYCLOAK_CLIENT_secret_")
```
in the implementation
```ruby
Rails.application.config.x.keycloak.client_d
Rails.application.config.x.keycloak.client_secret
```
and in the tests 

```ruby
allow(Rails.application.config.x).to receive(:keycloak).and_return(double(client_id: 123, client_secret: 123)
```

which prevents this PR from mutating the local `ENV`

Comment (StupidCodeFactory) [reply shared-web/app/clients/shared/web/token.rb]: and ditch this 
https://github.com/UKGovernmentBEIS/beis-opss-psd/pull/23/files#diff-7616f444028117bd2de8bf36babca2fdR6-R9

File: shared-web/spec/spec_helper.rb [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  # the `--only-failures` and `--next-failure` CLI options. We recommend
+  # you configure your source control system to ignore this file.
+  config.example_status_persistence_file_path = "spec/examples.txt"
+
+  # Limits the available syntax to the non-monkey patched syntax that is
+  # recommended. For more details, see:
+  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
+  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
+  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
+  config.disable_monkey_patching!
```
Comment (StupidCodeFactory) [inline shared-web/spec/spec_helper.rb]: you want to enable this one

File: shared-web/spec/spec_helper.rb [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  # order dependency and want to debug it, you can fix the order by providing
+  # the seed, which is printed after each run.
+  #     --seed 1234
+  config.order = :random
+
+  # Seed global randomization in this process using the `--seed` CLI option.
+  # Setting this allows you to use `--seed` to deterministically reproduce
+  # test failures related to randomization by passing the same `--seed` value
+  # as the one that triggered the failure.
+  Kernel.srand config.seed
```
Comment (StupidCodeFactory) [inline shared-web/spec/spec_helper.rb]: and this

File: shared-web/app/clients/shared/web/token.rb (hunk 7 above)
Comment (doctorpod) [reply shared-web/app/clients/shared/web/token.rb]: Can you be more explicit regarding the link at the end which points to differences in 12 files

**Comment by slorek** [comment]
Tested on review app and observed `/cases` page load taking ~2s on cache miss and 1s on cache hit, which appears to be a significant improvement.

**Comment by slorek** [comment]
Review app: https://psd-improved-perf-web.london.cloudapps.digital/

**Pull request merged**